### DISCUSSION TITLE
Hide terminal without nuking IO Pipe performance

### DISCUSSION THREAD

**Discussion by CramBL** [fltk-rs/fltk-rs#1463] [Q&A] [answered]
## TLDR: How can I hide the terminal from the user on Windows but still keep it open

Hi

So far I'm very happy with FLTK but I've run into a problem I cannot easily solve.

I'm building a cross-platform GUI that wraps a few other command-line utilities that I am also the author of. I don't want the user to be forced to be aware that I'm using a CLI in the background to handle a lot of utility, so naturally, I want to hide the terminal.

Through the very nice documentation, I found that I can hide the terminal in release mode with `#![cfg_attr(not(debug_assertions), windows_subsystem = "windows")]`, that's really nice, so far so good.

It just happens that it nukes the performance on Windows. 
It's nice and responsive in debug mode, but with the annoying empty terminal in the background, and in release mode simple things take almost 5 seconds. 

Now I assume that there's a lot of setting up/tearing down when the parent process (GUI) spawns children (CLIs) that need stdout/stdin/stderr and that's why I'm taking a hit on performance. 

So can I just hide the terminal?

I tried that and failed, and now I'm asking for help here 😄 

My failed attempt at hiding the windows terminal (from this SO https://stackoverflow.com/questions/29763647/how-to-make-a-program-that-does-not-display-the-console-window):
```rust
/// cargo.toml: 
/// winapi = {version = "0.3.9", features = ["wincon", "winuser"]}

fn hide_console_window() {
    use std::ptr;
    use winapi::um::wincon::GetConsoleWindow;
    use winapi::um::winuser::{ShowWindow, SW_HIDE};

    let window = unsafe { GetConsoleWindow() };
    // https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-showwindow
    if !window.is_null() {
        unsafe {
            ShowWindow(window, SW_HIDE);
        }
    }
}

fn hide_console_window_alt() {
    unsafe { winapi::um::wincon::FreeConsole() };
}
```

### EDIT:

For now, I do this:
```rust
#[cfg(windows)]
fn hide_console_window() {
    use winapi::um::wincon::GetConsoleWindow;
    use winapi::um::winuser;

    let window = unsafe { GetConsoleWindow() };
    if !window.is_null() {
        unsafe {
            // https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-showwindow
            // Hide takes no effect if the window is not active, so we activate it before hiding (minimizing) it.
            // The window being active at the time of calling this varies between launching the app.
            // If the window is not active when attempting to hide it, it will just be displayed behind the app.
            _ = winuser::ShowWindow(window, winuser::SW_SHOWMINIMIZED);
            let _was_visible = winuser::ShowWindow(window, winuser::SW_HIDE);
            // https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-enablewindow
            winuser::EnableWindow(window, 0);
        }
    }
}
```

It just minimizes the terminal window and deactivates mouse/keyboard input. Moving to another way of IPC seems to be the only reliable solution.

**Comment by MoAlyousef** [comment]
Hello

I'm not sure I understand the bit about setting up and tearing down the gui process. Are you spawing an FLTK window with every command?

Anyways, I assume for every command, you're spawning a `std::process::Command::new("some_cli_tool")` which might have an overhead.
You could spawn a single process at the start of your app, and that's the shell itself. Then in your callbacks spawn your cli tools from the spawned shell, and read the output in a separate thread. Something like:
```rust
#![cfg_attr(not(debug_assertions), windows_subsystem = "windows")]

use fltk::{prelude::*, *};
use std::io::Read;
use std::io::Write;
use std::process::{Command, Stdio};
use std::sync::{Arc, Mutex};

fn main() {
    // one time init Command
    let mut cmd = if cfg!(target_os = "windows") {
        Command::new("cmd.exe")
    } else if cfg!(target_os = "macos") {
        Command::new("/bin/sh")
    } else {
        let mut cmd = Command::new("/bin/bash");
        cmd.arg("-i");
        cmd
    };
    let mut child = cmd
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .stdin(Stdio::piped())
        .spawn()
        .unwrap();
    let writer = child.stdin.take().unwrap();
    let writer = Arc::new(Mutex::new(writer));
    let mut reader = child.stdout.take().unwrap();

    // init fltk
    let a = app::App::default();
    let mut buf = text::TextBuffer::default();
    let mut w = window::Window::default().with_size(400, 300);
    let mut col = group::Flex::default_fill().column();
    let row = group::Flex::default().row();
    col.fixed(&row, 30);
    let mut git_help = button::Button::default().with_label("git help");
    let mut ls_lah = button::Button::default().with_label("ls lah");
    row.end();
    let mut disp = text::TextDisplay::default();
    disp.set_buffer(buf.clone());
    col.end();
    w.end();
    w.show();

    git_help.set_callback({
        let writer = writer.clone();
        let mut buf = buf.clone();
        move |_b| {
            buf.set_text("");
            writer.lock().unwrap().write_all(b"git --help\n").unwrap();
        }
    });

    ls_lah.set_callback({
        let mut buf = buf.clone();
        move |_b| {
            buf.set_text("");
            writer.lock().unwrap().write_all(b"ls -lah\n").unwrap();
        }
    });

    std::thread::spawn(move || {
        while child.try_wait().is_ok() {
            let mut msg = [0u8; 1024];
            if let Ok(sz) = reader.read(&mut msg) {
                buf.append(&String::from_utf8(msg[0..sz].to_vec()).unwrap());
                app::awake();
            }
            std::thread::sleep(std::time::Duration::from_millis(30));
        }
    });

    a.run().unwrap();
}
```
I' ve only tested this on linux, but it should similarly work on windows.

**Reply by CramBL** [reply]
Thank you for the thorough reply. Your method will allow me to improve my current workaround a bit, but it does not solve the inherent issue. A terminal window is still spawned on Windows, and as you noted this does not happen on Linux, it is windows specific behaviour, unfortunately.

I'm considering named pipes as an alternative, but that will be a much more involved process which I had hoped to postpone until the app was in a much later stage.

> I'm not sure I understand the bit about setting up and tearing down the gui process. Are you spawing an FLTK window with every command?

I meant that setting up the terminal, as it appears that the commands run with `std::process:command` (when used with `#![windows_subsystem = "windows"]`) each spawn a new terminal, run the command, and then destroy it again. 

I also suspect that Windows Defender might be scanning it or slowing it down in some other way, I have observed that Windows Defender has blocked me from running the app while experimenting with solutions to the issue (something I have never experienced before with Rust or C++ projects).

**Reply by MoAlyousef** [reply, marked as answer]
It seems there are creation flags which can be used with Command on windows:

https://stackoverflow.com/questions/59692146/is-it-possible-to-use-the-standard-library-to-spawn-a-process-without-showing-th

I haven’t tried these yet since I’m away from my windows machine these few days.

**Reply by CramBL** [reply]
Yes! Thank you!

I modified your example a little bit and it worked.

> Added CREATE_NO_WINDOW constant
> Modified Windows branch of the expression that creates the shell

```rust
const CREATE_NO_WINDOW: u32 = 0x08000000;

let mut cmd = if cfg!(target_os = "windows") {
    let mut w_cmd = Command::new("cmd.exe");
    w_cmd.creation_flags(CREATE_NO_WINDOW);
    w_cmd
} else if cfg!(target_os = "macos") {
    Command::new("/bin/sh")
} else {
    let mut cmd = Command::new("/bin/bash");
    cmd.arg("-i");
    cmd
};
```

I still haven't tested the implementation for my app, but from my understanding, this solves the issue completely. It's just a much larger refactor required for my app so it will have to wait a bit.

**Reply by CramBL** [reply]
Ok final conclusion here for any future readers.

The solution that @MoAlyousef provided was almost what I needed but not quite, as it reads from a buffer of the output of the shell child process, and so querying command exit codes and just knowing exactly when a command finishes, is not obvious (at least I couldn't figure it out).

So what I ended up doing, is invoking the binaries of the other command-line utilities directly, and using a couple of utility functions for convenience, see below.

Then I didn't have to change anything else at all. Tested on Windows 11 and Lubuntu 22.04.3

```Rust
use std::{
    ffi::OsStr,
    process::{Command, Output, Stdio},
};

#[cfg(target_os = "windows")]
use std::os::windows::process::CommandExt;
#[cfg(target_os = "windows")]
const CREATE_NO_WINDOW: u32 = 0x08000000;

pub(super) fn run_with_args<I, S>(bin: S, args: I) -> Result<Output, std::io::Error>
where
    I: IntoIterator<Item = S>,
    S: AsRef<OsStr>,
{
    let mut cmd = Command::new(bin);

    #[cfg(target_os = "windows")]
    cmd.creation_flags(CREATE_NO_WINDOW);

    cmd.stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .stdin(Stdio::piped())
        .args(args);

    cmd.output()
}

pub(super) fn run_with_arg<S>(bin: S, arg: S) -> Result<Output, std::io::Error>
where
    S: AsRef<OsStr>,
{
    let mut cmd = Command::new(bin);

    #[cfg(target_os = "windows")]
    cmd.creation_flags(CREATE_NO_WINDOW);

    cmd.stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .stdin(Stdio::piped())
        .arg(arg);

    cmd.output()
}
```
